### PR TITLE
Support for CentOS Stream 9

### PR DESCRIPTION
**Pull request by phracek**
This pull request add support for testing, building and pushing s2i-php-container into Quay.io.

The CentOS Stream 9 image is pushed to quay.io/sclorg namespace.

@hhorak @pkubatrh @zmiklank WDYT?

### CODE REVIEW

**Comment by phracek** [comment]
[test-all]

**Comment by zmiklank** [comment]
@phracek What tests do you specifically mean?

A solution can be creating a new feature branch here in the upstream repository and merging PRs for tests and test support into it. It could be then merged into the master when everything works.
It seems, that GitHub actions tests aren't restricted to run on the master branch only, so that shouldn't be a problem.

**Comment by zmiklank** [comment]
> @zmiklank I have meant, first of all, merge .github/workflows and then run tests for the PR.
> Currently, when we schedule tests, then CentOS Stream 9 is skipped because it does not exist in the master branch.

That means, that this PR is to be split into 2 different ones if I understand correctly?

**Comment by phracek** [comment]
> That means, that this PR is to be split into 2 different ones if I understand correctly?

Yes, I will do it, if it makes sense.

**Comment by zmiklank** [comment]
Yes, in my opinion, it makes sense. Thanks.

**Comment by phracek** [comment]
GitHub Action PR is here: https://github.com/sclorg/s2i-php-container/pull/348. As soon as it is going to be merged, I will rebase this one.

**Comment by zmiklank** [comment]
[test-all]

**Comment by phracek** [comment]
Let's try first shot with testing CentOS Stream 9
[test]

**Comment by phracek** [comment]
Let's try first shot with testing CentOS Stream 9

[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by zmiklank** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
Difference between 7.4/Dockerfile.rhel8 and 8.0/Dockerfile.c9s for faster review:

```bash
diff 7.4/Dockerfile.rhel8 8.0/Dockerfile.c9s
1c1
< FROM ubi8/s2i-base:1
---
> FROM quay.io/sclorg/s2i-base-c9s:c9s
14,15c14,15
< ENV PHP_VERSION=7.4 \
<     PHP_VER_SHORT=74 \
---
> ENV PHP_VERSION=8.0 \
>     PHP_VER_SHORT=80 \
35c35
<       name="ubi8/${NAME}-${PHP_VER_SHORT}" \
---
>       name="sclorg/${NAME}-${PHP_VER_SHORT}-c9s" \
40c40
<       usage="s2i build https://github.com/sclorg/s2i-php-container.git --context-dir=${PHP_VERSION}/test/test-app ubi8/${NAME}-${PHP_VER_SHORT} sample-server" \
---
>       usage="s2i build https://github.com/sclorg/s2i-php-container.git --context-dir=${PHP_VERSION}/test/test-app sclorg/${NAME}-${PHP_VER_SHORT}-c9s sample-server" \
44,46c44,45
< RUN yum -y module enable php:$PHP_VERSION && \
<     INSTALL_PKGS="php php-mysqlnd php-pgsql php-bcmath \
<                   php-gd php-intl php-json php-ldap php-mbstring php-pdo \
---
> RUN INSTALL_PKGS="php php-mysqlnd php-pgsql php-bcmath \
>                   php-gd php-intl php-ldap php-mbstring php-pdo \
```

**Comment by phracek** [comment]
[test-all]

**Comment by phracek** [comment]
[test-all]

**Comment by phracek** [comment]
I have hit the issue with cakephp-ex https://github.com/sclorg/cakephp-ex/issues/130 with test `test_build_from_dockerfile`

**Comment by phracek** [comment]
[test-all]

**Comment by paulbadcock** [comment]
@phracek thanks for the hard work, so glad the work towards stream + FPM has started.

I've been testing with this and found a few issues.

1. php-fpm clears environment variables
  This will cause a lot of issues as environment variables will be needed to pass in config settings from the pod environment/configmaps/secrets to the app.
  
  Here's a commit that fixes this issue by just including the www.conf in the repo
  https://github.com/paulbadcock/s2i-php-container-1/commit/7959e0ea205083bdd4d23ea45ca7917bb4d2b22f
  
 2. Default folder location is no longer /web
    This will break any legacy upgrades to the container as the new environment variable will be needed for the `DOCUMENTROOT`. Not sure if its intended but it's a possible breakage users will run into.

3.  The default fpm config www.conf will cause ephemeral growth
   Not sure if echoing out the data to stdout would be better but packing in the www.conf might be the best way to allow the environment variables like `SESSION_PATH` to tweak the settings of the worker like in the older mod_php containers.
 
Let me know how I can help out or if this should be on a separate issue

**Comment by phracek** [comment]
@paulbadcock Thanks for you help. I will set `clear_env = no` in common.sh file, function 'config_general'

File: 8.0/root/usr/libexec/container-setup [hunk 1]
```diff
@@ -41,18 +41,19 @@ mkdir -p ${APP_ROOT}/etc
...
 chown -R 1001:0 ${HTTPD_DATA_PATH}
 chmod -R a+rwx ${PHP_SYSCONF_PATH}
-mkdir -p /run/php-fpm
-chmod -R a+rwx /run/php-fpm
-chown -R 1001:0 /run/php-fpm
-mkdir -p /var/log/php-fpm
-chmod -R a+rwx /var/log
-#chown -R 1001:0 /var/log/php-fpm
-
+if [ "x$PLATFORM" == "xel9" ]; then
```
Comment (remicollet) [inline 8.0/root/usr/libexec/container-setup:49]: Could be better to use same test (`[ -d "/run/php-fpm" ]`) everywhere for consistency ?

Comment (remicollet) [reply 8.0/root/usr/libexec/container-setup:49]: notice: especially as php-fpm can also be used (if wanted) on EL-8

File: 8.0/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,80 @@
+FROM quay.io/centos7/s2i-base-centos7
```
Comment (hhorak) [inline 8.0/Dockerfile:1]: We should not add a Dockerfile for RHSCL, which do not have php 8.0.

File: 8.0/Dockerfile.fedora [hunk 3]
```diff
@@ -0,0 +1,73 @@
+FROM registry.fedoraproject.org/f32/s2i-base:latest
```
Comment (hhorak) [inline 8.0/Dockerfile.fedora:1]: Moving to f35, we can use 8.0 in Fedora as well. It may be done in a separate PR though, as this one focuses on C9S.

File: 8.0/Dockerfile.rhel8 [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+EXPOSE 8080
+EXPOSE 8443
+
+# Description
+# This image provides an Apache 2.4 + PHP 7.4 environment for running PHP applications.
+# Exposed ports:
+# * 8080 - alternative port for http
+
+ENV PHP_VERSION=7.4 \
```
Comment (hhorak) [inline 8.0/Dockerfile.rhel8:14]: Old version, likely something to be changed later in a separate PR.

File: 8.0/README.md [hunk 5]
```diff
@@ -0,0 +1,330 @@
+PHP 8.0 container image
+=======================
+
+This container image includes PHP 8.0 as a [S2I](https://github.com/openshift/source-to-image) base image for your PHP 8.0 applications.
+Users can choose between RHEL and CentOS based builder images.
```
Comment (hhorak) [inline 8.0/README.md:5]: We can say CentOS Stream specifically, to not confuse readers.

File: 8.0/root/opt/app-root/etc/scl_enable [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# IMPORTANT: Do not add more content to this file unless you know what you are
+#            doing. This file is sourced everytime the shell session is opened.
+#
+# This will make scl collection binaries work out of box.
+unset BASH_ENV PROMPT_COMMAND ENV
+source scl_source enable ${SCL_ENABLED} httpd24 $NODEJS_SCL
```
Comment (hhorak) [inline 8.0/root/opt/app-root/etc/scl_enable:6]: This file can be removed as we do not have SCL variant for php 8.0.

File: 8.0/root/usr/libexec/container-setup (hunk 1 above)
Comment (phracek) [reply 8.0/root/usr/libexec/container-setup:49]: Do we want to support PHP-8.0 also for RHEL8 containers? I would prefer to strictly use it only in CentOS Stream 9.

File: 8.0/root/opt/app-root/etc/scl_enable (hunk 6 above)
Comment (phracek) [reply 8.0/root/opt/app-root/etc/scl_enable]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/b2bfb6c3f2fe144f39784203075ac243a0216f72

File: 8.0/cccp.yml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# This is for the purpose of building this container
+# on the CentOS Container Pipeline.
+job-id: php-74-centos7
```
[the remark this answers is not in the document]
Comment (phracek) [reply 8.0/cccp.yml]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/b2bfb6c3f2fe144f39784203075ac243a0216f72

File: 8.0/README.md (hunk 5 above)
Comment (phracek) [reply 8.0/README.md]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/8f076621afa4c74b55f27a51f365b8824a934f6b

File: 8.0/Dockerfile.rhel8 (hunk 4 above)
Comment (phracek) [reply 8.0/Dockerfile.rhel8]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/b2bfb6c3f2fe144f39784203075ac243a0216f72

File: 8.0/Dockerfile (hunk 2 above)
Comment (phracek) [reply 8.0/Dockerfile]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/b2bfb6c3f2fe144f39784203075ac243a0216f72

File: 8.0/Dockerfile.c9s [hunk 8]
```diff
@@ -41,7 +41,7 @@ LABEL summary="${SUMMARY}" \
       maintainer="SoftwareCollections.org <sclorg@redhat.com>"
 
 # Install Apache httpd and PHP
-RUN INSTALL_PKGS="php php-mysqlnd php-pgsql php-bcmath \
+RUN INSTALL_PKGS="php php-fpm php-mysqlnd php-fpm php-pgsql php-bcmath \
```
[the remark this answers is not in the document]
Comment (phracek) [reply 8.0/Dockerfile.c9s]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/a23b75205794bcc388e36e718651e6a1d9dbb0e3

File: 8.0/root/usr/libexec/container-setup (hunk 1 above)
Comment (phracek) [reply 8.0/root/usr/libexec/container-setup]: Addressed by https://github.com/sclorg/s2i-php-container/pull/347/commits/a839d56b3aa873bd226fea6082b0f2d0eeee26fc. As condition for CentOS Stream 9 as check for existence directory `/run/php-fpm` is present.

Comment (phracek) [reply 8.0/root/usr/libexec/container-setup]: The question is if we would like to add PHP-FPM also in Dockerfile.rhel8, which is now present, or we would like to use only `mod_php` like in `7.4/Dockerfile.rhel8` https://github.com/sclorg/s2i-php-container/blob/master/7.4/root/usr/share/container-scripts/php/common.sh#L29. Nowadays I would prefer for Dockerfile.rhel8 use mod_php. Like before.

**Comment by phracek** [comment]
> @phracek thanks for the hard work, so glad the work towards stream + FPM has started.
> 

> Here's a commit that fixes this issue by just including the [www.conf](http://www.conf) in the repo [paulbadcock@7959e0e](https://github.com/paulbadcock/s2i-php-container-1/commit/7959e0ea205083bdd4d23ea45ca7917bb4d2b22f)

This is addressed by commit https://github.com/sclorg/s2i-php-container/pull/347/commits/66f052a0b2f7d159b90c65bcaa8e327607cbdbea

**Comment by phracek** [comment]
[test-all]

**Comment by pkubatrh** [comment]
[test]

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
[test-openshift]

**Comment by hhorak** [comment]
logs for centos9 and fedora are not available any more, re-running because that are the only OS version we have 8.0.

**Comment by hhorak** [comment]
[test]

**Comment by hhorak** [comment]
LGTM
